c/misra/src/rules/RULE-8-15/RedeclarationOfObjectWithoutAlignment.ql

```diff
@@ -0,0 +1,97 @@
+/**
+ * @id c/misra/redeclaration-of-object-without-alignment
+ * @name RULE-8-15: Alignment should match between all declarations of an object
+ * @description An object declared with an explicit alignment shall be explicitly aligned in all
+ *              declarations.
+ * @kind problem
+ * @precision very-high
+ * @problem.severity error
+ * @tags external/misra/id/rule-8-15
+ *       external/misra/c/2012/amendment3
+ *       readability
+ *       maintainability
+ *       external/misra/obligation/required
+ */
+
+import cpp
+import codingstandards.c.misra
+
+/**
+ * Performance optimization; start query by joining attributes to declarations
+ * rather than locations.
+ *
+ * Including the entry location also speeds up search.
+ */
+newtype TAttributeDeclLocation =
+  TAttributeDeclLocationInfo(Attribute attribute, DeclarationEntry entry, Location entryLocation) {
+    entry.getDeclaration().(Variable).getAnAttribute() = attribute and
+    entryLocation = entry.getLocation()
+  }
+
+/**
+ * Get a DeclarationEntry along with its explicitly declared Attributes.
+ *
+ * DeclarationEntry does not have a method for getting Attributes by default,
+ * because an attribute declared on any DeclarationEntry affects all others,
+ * and attributes really belong to the declared variable rather than the
+ * declaration itself.
+ *
+ * In order to support this rule, we find for each attribute
+ * - A declaration entry which
+ * - corresponds to a variable associated with this attribute
+ * - is in the same file as this attribute
+ * - has identifier location after the attribute declaration
+ * - has no other declaration entry between this one and the attribute.
+ *
+ * This should give us a highly reliable means of finding which attributes are
+ * associated with which `DeclarationEntry`s.
+ *
+ * One note of caution: the associated `Variable` must be treated with caution,
+ * as there are multiple instances of that `Variable` if it is declared
+ * multiple times, they equal each other, and `getLocation()` on each variable
+ * returns every location result. This class must act on `DeclarationEntry`s to
+ * deliver reliable results.
```

Comment:
In general, we should see a single instance of `Variable` for the same variable in the program, and the location of that variable should be the location of the definition(s). There are some exceptions to this: if a variable is declared in multiple link targets which are never statically linked in the program, or if a file is recompiled under a different set of assumptions, which can created - but in general those should not be considered here. I would therefore reword this paragraph for clarity. Note: it's certainly true that this class needs to consider `DeclarationEntry`'s, though.